### DISCUSSION TITLE
Double Take Motion Integration

### DISCUSSION THREAD

**Discussion by MooreSi** [tsightler/ring-mqtt#501] [Support Q&A]
Hoping for some assistance/guidance, thanks first of all for ring-mqtt as this is a great app that works to integrate and expose my Ring Doorbell Pro 2. I can live stream directly into Frigate and then use Frigate within Double Take as a source but this isn't elegant and has issues such as my doorbell fails to work and my Alexa's do not alert on motion as I suspect the stream/traffic is busy communicating to the AWS/Ring servers and Frigate. Using MQTT Explorer I can see the topics so my question is can anyone provide guidance on how to add code into the Double Take config.yaml that imports a number of high-res images when it detects motion or possibly only streams for a small period of time, I'm then using Compreface to facial detection which works fine with my Hikvision cameras via Frigate but these are either side of the house and I'd like to use my Ring doorbell as people approach the front door. Running HAAS on VirtualBox/Debian, ring-mqtt, Frigate and CompreFace on Docker, also have Coral USB.

Thanks in advance

**Comment by tsightler** [comment]
As stated in the documentation, use of ring-mqtt with Frigate and other NVR systems is not a supported use case for this project so I will lock and delete this thread in 24 hours.

**Reply by MooreSi** [reply]
Accept that and looking for guidance of pulling images or video upon motion detection from the Ring doorbell not using Frigate but unsure how to. I've read through a lot of docs but can't get it to work just yet.

KR

**Reply by tsightler** [reply]
The simplest thing is to use an automation which runs a shell script that calls ffmpeg, from which you can create whatever output you like.  You could start with something simple like:

> ffmpeg -rtsp_transport tcp -y -i rtsp://<ha-instance-ip>:8554/xxxxxx_live -frames:v 1 -ss 00:00:01 /media/image.jpg

Run that with an automation and it should capture a single jpg image from early in the stream, but then you can expand on that to copy video to somewhere else, serve it up via RTSP for X number of seconds, or whatever you like.

The biggest problem with all of these NVR systems is that, as far as I can tell, all of them expect a 24x7 camera, and that's just not what a Ring camera is designed for.  IMO, people should push the NVR tools to be more friendly to event driven cameras, if they implemented a simple way to enable/disable capture based on an external event then this would all be super easy.

**Reply by MooreSi** [reply]
Thank you very much I shall try that, I'd ruled out using Frigate or any type of NVR but was looking for an option to somehow capture image(s) of people close to the front door to then run facial recognition against. 

Cheers

**Reply by tsightler** [reply]
I've been trying to figure out a way to "stream forever", basically just have a stream that is the snapshot but then replace it with the live stream whenever there is actual activity, but every method I've found just adds a lot of complication to the existing codebase and isn't very reliable.  Someone with more streaming skills could probably solve that problem.

**Reply by tsightler** [reply]
Spent a little time browsing the Frigate Github page and found that there are several feature requests asking for better compatibility with cameras that are not designed for continuous streaming so maybe this will happen at some point.

**Reply by MooreSi** [reply]
Thanks for looking further into this, I’ll have a play around with it at the weekend and see how far I get. I’ll post the results here if successful

**Reply by tsightler** [reply]
OK, I'll leave this thread available for now in case you get some results.

**Reply by MooreSi** [reply]
Sort of got something working but unsure if either MQTT or ring-mqtt doesn't snapshot frequently enough and it often misses people but its a start. The below is added to the Double Take, config.yml. It may be the case that its possible by creating an automation that on detecting motion takes a series of snapshots/images which are then ingested into the detector using Double Take, that's my next step to see if images can be directly loaded in

cameras:
#  mqtt:
    doorbell:
      snapshot:
        topic: ring/xxxxxxxxxxxxxx/camera/54e01988cf84/snapshot/image
        url: /media/Ring/ring_snapshot_automation2.jpg

**Reply by 503Dev** [reply]
Just for anyone else looking into this: These cameras really do not play nice with 24x7 streaming.  I am not saying you can't do it but I have both wired and battery powered models.  The battery powered will drain their battery in 48 hours if you try to stream frequently.

The wired will work but I noticed they start to get laggy and drop packets after a few days with non-stop streaming.  I don't think the onboard hardware is designed for this and that is the main issue.

**Reply by tsightler** [reply]
I have warnings all over this project about this, but people always want to try it for themselves.

**Comment by leccelecce** [comment]
Hi @tsightler - thank you for your truly excellent library. I have been spending a fair bit of time using ring-mqtt, Node-Red and Double-Take to try and do live facial recognition off my solar-charged Battery Doorbell 3, which then sends on notifications to Home Assistant, as described in the previous comment. I'm not attempting to stream continuously from the doorbell, but rather use the motion state = ON or ding state = ON as the trigger to begin streaming. What I've ended up with looks like:

`/usr/bin/ffmpeg -rtsp_transport tcp -i rtsp://ring-mqtt:8554/xxxxxxxxxxxxx_live -an -vframes 5 -qscale:v 2 -vf fps=1 -y /data/snapshots/doorbell/img%03d.jpg`

...with the jpgs then being served to Double-Take in realtime as they're written for image recognition.

TL;DR: I would like to look into sending you a PR that extends ring-mqtt (probably via borrowing from livecall.js) to provide a function to generate JPGs from the ring-client-api ffmpeg stream directly, then sticking the binary JPG data on to a new mqtt topic directly (similar for how you do for snapshots on the /snapshot/image topic). I would probably make frame rate to save, and number of seconds to run for configurable.

Benefits of doing this for my use case: snapshots would be full-resolution, could avoid going via rtsp-simple-server, and could avoid me having to write them out to disk only to then serve them up to Double-Take manually over http.

I'm aware that this isn't very efficient due to the round-trip to Ring's servers plus battery use of pulling a video stream just to chop out some JPG frames, but I was wondering if you'd consider such a PR or prefer not to have that implemented?

**Reply by tsightler** [reply]
I had previously played around with a similar idea as, before push notifications, the only way to get a "snapshot" image from a battery camera on motion was to grab one from the stream so the code actually started a stream and grabbed an image.  I had hoped that maybe I could pipe a low framerate video via mqtt, but I found it to be able to handle basically just a frame or two every 3-4 seconds at best although this may have been due to the HA implementation of the snapshot camera more than anything.  I've found a few posts where people were able to get 1 fps via mqtt using other technology, and at least one example of 5 fps.

However, there was a much bigger issue with this approach, starting a stream is the same as answering a motion notification on the Ring app, so that meant that every motion event displayed as "answered", not really a major issue, but one that a lot of users complained about.  The bigger issue was that, because it was an "answer" this meant that recordings stopped as soon as the stream ended, so, even if I just wanted a single frame for a snapshot, I had to keep the stream running or it would truncate the recording immediately instead of recording for the set duration in the app.   I compensated for this by reading the configured recording duration, and keeping the stream alive even when I didn't need it, but it was trouble prone.

I'm not super interested in going down this path again for a case that I don't really even want people doing with ring-mqtt, although, I'm not going to lie, I've actually been tinkering with what is a similar idea.  My current plan is to replace rtsp-simple-server with go2rtc in the next significant update to ring-mqtt.  Most of this is due the fact that rtsp-simple-server really struggles with Ring transcoded files and I really want to provide that as an option vs the raw falls as many have asked about video with pre-roll and timestamps, and it would also allow HA uses to leverage WebRTC with ring-mqtt for low-latency viewing without having to run the separate go2rtc addon.

However, those aren't the only reasons.  The go2rtc server also has the ability to convert a series of snapshots directly into an MJPEG stream.  My thought here is that I could upscale the snapshot images and feed the standard snapshot images to this stream, however, on a detected motion, I could then pipe a 5-10 FPS stream of MJPEG images uses basically the method you describe, any time a motion is detected.  This would allow for tools that need a continous stream to get one, but it would normally just be a low frequency snapshot image, while, during any motion event, there would be enough frames to actually detect motion properly.

Now, this is just an idea for now, I haven't had much time to play with this and honestly don't know when I might get to it.  I'm also waiting on one key bug to be fixed in go2rtc.  My current thought is to start with just replacing rtsp-simple-server with go2rtc without breaking anything, and then, after that work on the "snapshot stream" idea.  This would be enabled with an option.

**Reply by leccelecce** [reply]
Thank you for the prompt and detailed reply!

> I had hoped that maybe I could pipe a low framerate video via mqtt, but I found it to be able to handle basically just a frame or two every 3-4 seconds at best although this may have been due to the HA implementation of the snapshot camera more than anything. I've found a few posts where people were able to get 1 fps via mqtt using other technology, and at least one example of 5 fps.

I haven't observed any problems here - I can quite comfortably pump 5 fps through ring-mqtt->mosquitto->node-red->double-take. But I'm not trying to view any video in HA so it's quite possibly a limitation there as you say.



> However, there was a much bigger issue with this approach, starting a stream is the same as answering a motion notification on the Ring app, so that meant that every motion event displayed as "answered"

I hadn't spotted this until you pointed it out. It doesn't bother me personally, but I can see it's not ideal.


> Most of this is due the fact that rtsp-simple-server really struggles with Ring transcoded files

Interested as to what you're referring to here? It looks to me like it works pretty reliably, and with minimal CPU requirement. The one thing that's frustrating me at the minute is the ~6 seconds it takes for a stream to start flowing to ffmpeg. I've been playing around with ffmpeg options as it seems like it needs a certain amount of the stream to confirm the parameters, but adjusting this down doesn't help much. From lining up the ring-mqtt and ffmpeg logs it looks like you report the stream as active within < 1 second. Would you expect go2rtc to improve on this?

> The go2rtc server also has the ability to convert a series of snapshots directly into an MJPEG stream.

So that's something that is presumably viable for "high power" cameras that can snapshot during motion events, but not battery ones as I understand it?

**Reply by tsightler** [reply]
> I haven't observed any problems here - I can quite comfortably pump 5 fps through ring-mqtt->mosquitto->node-red->double-take. But I'm not trying to view any video in HA so it's quite possibly a limitation there as you say.

It could have possibly been the MQTT.js library itself, as that code has some serious issues with larger MQTT messages back when I first started using it, including some memory leaks, but those were fixed up in the last year.  I honestly haven't tried since then.

> I hadn't spotted this until you pointed it out. It doesn't bother me personally, but I can see it's not ideal.

It wasn't so much that the "answer" was an issue itself, it was that, if you only kept the stream alive for like 5 seconds, it truncated the recording to 5 seconds even if you have the default recording length set long.  Users really didn't like this.  That mean, for every single motion/ding event, I had to keep the stream alive for 30 or 60 seconds, even though I really only needed the first keyframe.  For users that didn't have a subscription it might have been less of an issue.

Regarding rtsp-simple-server, the videos from Ring cameras are not particularly clean.  This isn't unusual, most cameras make all kinds of compromises in their encoding to get decent video out of these fairly low-CPU powered devices, but tools like FFmpeg and VLC have had decades of dealing with all of these poor videos that they will usually play them with little to no visble artifacts.  However, rtsp-simple-server doesn't have as many hacks for these various issues in the H.264 encoding, and it struggles with some of those short cuts, for example when Ring transcodes the stream the splice between pre-roll and standard video is "rough", and rtsp-simple-server will stutter.  There are also not very many keyframes in the stream, and rtsp-simple-server sometimes seems to miss the first keyframe, so it can take many seconds for a client attached to the stream to actually join the stream.

Now, don't get me wrong, overall rtsp-simple-server is good, but I've tested playing back a local file directly via VLC, then streaming that same file via RTSP and the identical FFmpeg command via rtsp-simple-server and go2rtc, and, with rtsp-simple-server you get stutters and delayed playback, while with go2rtc it's basically indistinguishable from playing the file directly.

Now, whether it would cut down on the 6 seconds, probably not, maybe a tad when keyframe caching is implemented.  Honestly, I've worked hard to get the streaming startup time down, but there's only so much that can be done.  In my setup my streams start in ~2.5 seconds, while in the Ring app it's about 1 second, but that's not really a fair comparison as the Ring app using native WebRTC directly to Ring server while in my case you have a client connect to RTSP, which sends MQTT message to start video, which then negotiates the WebRTC connection.  I report the stream "active" as soon as the WebRTC reaches accepted state, which is usually less than a second, but the actual stream still has to travel through werift (WebRTC implemented entirely in Typescript so decidedly not native), be piped to ffmpeg, out to rtsp-simple-server, and finally to a client.

I've tried all kinds of tuning with ffmpeg, but I can't really get the startup time down.  It's mostly about available bandwidth between the camera, the internet and the latency between all of that.  Low-power devices also run their wireless in a power-saving mode, which increases the initial latency (I don't own a single "low-power" Ring device, they just aren't for me).

TL;DR: Using go2rtc might slightly decrease startup time, but I don't expect it to be too much difference, however, I do believe it would provide a stream with less stutter during startup and especially for recorded/transcoded streams.

> So that's something that is presumably viable for "high power" cameras that can snapshot during motion events, but not battery ones as I understand it?

It's potentially usable for all cameras since it would just start a stream and forward low framerate mjpeg after a motion event directly from the stream.
